snuba/utils/streams/consumers/backends/synchronized.py

```diff
@@ -0,0 +1,309 @@
+        on_assign: Optional[Callable[[Mapping[TStream, TOffset]], None]] = None,
+        on_revoke: Optional[Callable[[Sequence[TStream]], None]] = None,
+    ) -> None:
+        # Update the subscription, so that the commit log consumer can starts
+        # to consume the commit log data for the subscribed (and potentially
+        # assigned) topics.
+        self.__subscription.set(Subscription(topics))
+
+        # TODO: How does the Kafka consumer handle paused streams? My
+        # assumption here is that the new subscription should cause those
+        # streams to be resumed again, but not sure that's accurate.
+        self.__paused_streams.clear()
+
+        def assignment_callback(streams: Mapping[TStream, TOffset]) -> None:
+            with self.__subscription.get() as subscription:
+                for stream, offset in streams.items():
+                    remote_offset = subscription.get_effective_remote_consumer_group_offset(
+                        stream, self.__remote_consumer_groups
+                    )
+                    if remote_offset is None or offset >= remote_offset:
+                        self.__backend.pause([stream])
+                    elif stream not in self.__paused_streams:
+                        self.__backend.resume([stream])
+
+            if on_assign is not None:
+                on_assign(streams)
+
+        return self.__backend.subscribe(
+            topics, on_assign=assignment_callback, on_revoke=on_revoke
+        )
+
+    def unsubscribe(self) -> None:
+        # Update the subscription, so that the commit log consumer can stop
+        # consuming the commit log data for the previous (and no longer
+        # potentially assigned) topics.
+        self.__subscription.set(Subscription([]))
+
+        # TODO: How does the Kafka consumer handle paused streams? My
+        # assumption here is that the new subscription should cause those
+        # streams to be resumed again, but not sure that's accurate.
+        self.__paused_streams.clear()
+
+        return self.__backend.unsubscribe()
+
+    def poll(
+        self, timeout: Optional[float] = None
+    ) -> Optional[Message[TStream, TOffset, TValue]]:
+        # Check to see if any streams are paused that are able to be resumed
+        # based on the progression of the remote offset. (This could go the
+        # other way as well -- remote offsets may be rolled back -- but those
+        # will be caught later during normal consumption. The only benefit to
+        # pausing them here would be to avoid buffering messages that will
+        # likely be discarded. On the other hand, by the time we get to consume
+        # these messages, the offsets may have been synchronized at that point.
+        # It's hard to predict which scenario is more likely -- in the interest
+        # of buffering and low latency, it probably makes sense to avoid
+        # pausing here?)
+        with self.__subscription.get() as subscription:
```

Comment:
Does all this block need to be synchronized considering you are not writing on the state ?

Reply:
Yes, since the other thread may be trying to mutate this data structure at the same time as we were reading it, we wouldn't get a consistent view of the data structure.

It's possible some of this might be able to be made a bit more efficient with a reader/writer lock instead of a single lock, but my hunch is that contention here is primarily going to be on one thread reading (the local consumer) and one thread writing (the commit log consumer.)

Reply:
ok, yes being on the safe side synchronizing the whole block seems the right thing unless we have evidence of it being a perf issue.